src/witness_generator/machines/fixed_lookup_machine.rs

```diff
@@ -76,60 +74,82 @@ impl FixedLookup {
-            Err(err) => Err(format!("First expression on the LHS is unknown: {err}")),
-        }?;
-
-        let right_key = right.expressions.first().unwrap();
-        let rhs_row = if let Expression::PolynomialReference(poly) = right_key {
-            // TODO we really need a search index on this.
-            fixed_data.fixed_cols
-                .get(poly.name.as_str())
-                .and_then(|values| values.iter().position(|v| *v == left_key))
-                .ok_or_else(|| {
-                    format!(
-                        "Unable to find matching row on the RHS where the first element is {left_key} - only fixed columns supported there."
-                    )
-                })
-                .map(|i| i as DegreeType)
-        } else {
-            Err("First item on the RHS must be a polynomial reference.".to_string())
-        }?;
+        let columns = right
+            .expressions
+            .iter()
+            .map(|right_key| {
+                if let Expression::PolynomialReference(poly) = right_key {
+                    fixed_data
+                        .fixed_cols
+                        .get(poly.name.as_str())
+                        .ok_or_else(|| {
+                            format!(
+                                "Unable to find a fixed column for the fixed lookup: {}",
+                                poly.name
+                            )
+                        })
+                } else {
+                    Err("RHS must be a polynomial reference.".to_string())
+                }
+            })
+            .collect::<Result<Vec<_>, _>>()?;
 
-        // TODO we only support the following case:
-        // - The first component on the LHS has to be known
-        // - The first component on the RHS has to be a direct fixed column reference
-        // - The first match of those uniquely determines the rest of the RHS.
+        let len = columns[0].len();
 
-        // TODO in the other cases, we could at least return some improved bit constraints.
+        let res = (0..len)
+            // get all lookup rows which match the lhs
+            .filter_map(|row| {
+                left.iter()
+                    .zip(columns.iter())
+                    .map(|(left, right)| {
+                        let right = &right[row];
+                        match left {
+                            Ok(left) => left
+                                .constant_value()
+                                // if the lhs is constant, it's a match iff the values match
+                                .map(|left| (left == *right).then_some(right))
+                                // if it's not constant, it's a match
+                                .unwrap_or_else(|| Some(right)),
+                            // if we do not know the lhs, it's a match
+                            Err(_) => Some(right),
+                        }
+                    })
+                    .collect::<Option<Vec<_>>>()
```

Comment:
I don't fully understand this. The `map` function above here returns an iterator with `Item = Option<AbstractNumberType>` and this is collected so that we get `None` whenever one of the items is None?